File: ios/Runner/FcmService.m
#import "FcmService.h"

@import UserNotifications;

@implementation FcmService

- (void)application:(UIApplication *)application
    didReceiveRemoteNotification:(NSDictionary *)userInfo {
    NSString *readNotificationId = [userInfo objectForKey:@"read_notification_id"];
    if (readNotificationId != nil) {
        [self onNotificationRead:readNotificationId];
    }
}

- (void)onNotificationRead:(NSString *)which {
    BOOL isAll = [which isEqualToString:@"all"];
    UNUserNotificationCenter *unc = [UNUserNotificationCenter currentNotificationCenter];
    
    [unc getDeliveredNotificationsWithCompletionHandler:^(NSArray<UNNotification *> *notifications) {
        int count = 0;
        for (UNNotification* notification in notifications) {
            NSDictionary* userInfo = notification.request.content.userInfo;
            
            NSString *notificationId = [userInfo objectForKey:@"notification_id"];
            if (notificationId == nil) {
                continue;
            }
            
            if (isAll || [which isEqualToString:notificationId]) {
                [unc removeDeliveredNotificationsWithIdentifiers:@[notification.request.identifier]];
                count++;
            }
        }
        
        NSLog(@"onNotificationRead(which=%@) dismissed=%d", which, count);
    }];
}

@end

File: ios/Runner/FcmService.swift
import UserNotifications

class FcmService {
  func onApplicationDidReceiveRemoteNotification(_ userInfo: [AnyHashable : Any]) {
    guard let readNotificationId = userInfo["read_notification_id"] as? String else { return }
    onNotificationRead(readNotificationId)
  }

  private func onNotificationRead(_ which: String) {
    let isAll = which == "all"
    let unc = UNUserNotificationCenter.current()

    unc.getDeliveredNotifications { notifications in
      var identifiers: [String] = []

      for notification in notifications {
        let userInfo = notification.request.content.userInfo
        guard let notificationId = userInfo["notification_id"] as? String else { continue }
        if isAll || notificationId == which {
          identifiers.append(notification.request.identifier)
        }
      }

      if !identifiers.isEmpty {
        unc.removeDeliveredNotifications(withIdentifiers: identifiers)
      }

      // TODO: switch to Logger when iOS 14 is available
      NSLog("onNotificationRead(which=\(which)) dismissed=\(identifiers.count)")
    }
  }
}
